core/src/docs/rfcs/2133_append_api.md

```diff
@@ -0,0 +1,106 @@
+- Proposal Name: `append_api`
+- Start Date: 2023-04-26
+- RFC PR: [apache/incubator-opendal#2133](https://github.com/apache/incubator-opendal/pull/2133)
+- Tracking Issue: 
+
+# Summary
+
+Introduce append operations for OpenDAL which allow users to add data to a file. And write operation will be not reentrant.
+
+# Motivation
+
+OpenDAL has the write operation used to create a file and append data to it. This is implemented based on multipart API. However, current approach has some limitations:
+
+- Data could be lost and not readable before w.close() returned Ok(())
+- File can't be appended again after w.close() returned Ok(())
+
+To address these issues, I propose adding an append operation and modifying the write operation to verify consistency between content length and data size. The revised write operation will not be reentrant; however, users can create an appender that provides a reentrant append operation.
+
+# Guide-level explanation
+
+The files created by the write operation can not be modified, only overwritten.
+
+```rust
+async fn write_test(op: Operation) -> Result<()> {
+    let bs = read_from_file();
+    op.write("path_to_file", bs).await?;
+  
+    // The following code will report an error
+    op.write("path_to_file", bs).await?;
+  
+    // Unless specified can overwrite
+    let option = OpWrite::default().with_allow_overwrite(true);
+    // content length must be the same as the data size
+    option.with_content_length(bs.len() as u64);
+    op.write_with("path_to_file", option, bs).await?;
+}
+```
+
+The files created by the append operation can be appended via multipart API or append API.
+
+```rust
+async fn append_test(op: Operation) -> Result<()> {
+  // create writer
+  let appender = op.appender("path_to_file").await?;
+
+  let bs = read_from_file();
+  appender.append(bs).await?;
+
+  let bs = read_from_another_file();
+  appender.append(bs).await?;
+
+  // close the file
+  // for multipart API, complete the upload and prevent writing after close
+  // for native append API, this is a no-op
+  appender.close().await?;
+
+  // abort the operation and delete all uploaded data
+  appender.abort().await?;
+}
+```
+
+# Reference-level explanation
+
+To implement this feature, we need to add a new API `append` into `oio::Append`. And remove useless function in `oio::Write`.
```

Comment:
No, this understanding is wrong.

We should not touch exsting APIs of `oio::Write`. They are different APIs.